File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        int [] frequency =  new int [2001];
        Map <Integer, List<Integer>> repList = new HashMap<>();
        int result [] = new int [k];
        int found = 0;

        for (int curr : nums) {
            frequency[curr + 1000]++;
        }

        for (int i = 0; i < 2001; i++) {
            int reps = frequency[i];
            repList.putIfAbsent(reps, new ArrayList<Integer>());
            repList.get(reps).add(i - 1000);
        }

        for (int i = nums.length; found < k; i--) {
            if (repList.containsKey(i)) {
                for (int j : repList.get(i)) {
                    result[found++] = j;
                }
            }
        }

        return result;
    }
}

File: Data Structures & Algorithms/count-subsequences/submission-0.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        prev = [0] * (len(t) + 1)
        curr = [0] * (len(t) + 1)
        prev[-1] = curr[-1] = 1

        for i in range(len(s) - 1, -1, -1):
            for j in range(len(t)):
                curr[j] = prev[j]
                if s[i] == t[j]:
                    curr[j] += prev[j + 1]
            prev, curr = curr, prev
        return prev[0]